packages/chevrotain/src/parse/grammar/lookahead_public.ts

```diff
@@ -0,0 +1,160 @@
+import {
+  ILookaheadStrategy,
+  ILookaheadValidationError,
+  IOrAlt,
+  LookaheadSequence,
+  LookaheadProductionType,
+  Rule,
+  TokenType,
+  OptionalProductionType
+} from "@chevrotain/types"
+import flatMap from "lodash/flatMap"
+import isEmpty from "lodash/isEmpty"
+import { defaultGrammarValidatorErrorProvider } from "../errors_public"
+import {
+  validateAmbiguousAlternationAlternatives,
+  validateEmptyOrAlternative,
+  validateNoLeftRecursion,
+  validateSomeNonEmptyLookaheadPath
+} from "./checks"
+import {
+  buildAlternativesLookAheadFunc,
+  buildLookaheadFuncForOptionalProd,
+  buildLookaheadFuncForOr,
+  buildSingleAlternativeLookaheadFunction,
+  getLookaheadPathsForOptionalProd,
+  getLookaheadPathsForOr,
+  getProdType,
+  PROD_TYPE
+} from "./lookahead"
+import { IParserDefinitionError } from "./types"
+
+export function getLookaheadPaths(options: {
+  occurrence: number
+  rule: Rule
+  prodType: LookaheadProductionType
+  maxLookahead: number
+}): LookaheadSequence[] {
+  const { occurrence, rule, prodType, maxLookahead } = options
+  const type = getProdType(prodType)
+  if (type === PROD_TYPE.ALTERNATION) {
+    return getLookaheadPathsForOr(occurrence, rule, maxLookahead)
+  } else {
+    return getLookaheadPathsForOptionalProd(
+      occurrence,
+      rule,
+      type,
+      maxLookahead
+    )
+  }
+}
+
+export class LLkLookaheadStrategy implements ILookaheadStrategy {
+  validate(options: {
+    rules: Rule[]
+    maxLookahead: number
+    tokenTypes: TokenType[]
+    grammarName: string
+  }): ILookaheadValidationError[] {
+    const leftRecursionErrors = this.validateNoLeftRecursion(options.rules)
+
+    let emptyAltErrors: IParserDefinitionError[] = []
+    let ambiguousAltsErrors: IParserDefinitionError[] = []
+    let emptyRepetitionErrors: IParserDefinitionError[] = []
+
+    if (isEmpty(leftRecursionErrors)) {
+      emptyAltErrors = this.validateEmptyOrAlternatives(options.rules)
+      ambiguousAltsErrors = this.validateAmbiguousAlternationAlternatives(
+        options.rules,
+        options.maxLookahead
+      )
+      emptyRepetitionErrors = this.validateSomeNonEmptyLookaheadPath(
+        options.rules,
+        options.maxLookahead
+      )
+    }
+
+    return leftRecursionErrors.concat(
```

Comment:
This is probably copied from the old code :)
But could we clean it up slightly and modernize it with:
- `let` --> `const`
- `[].concat([])` --> `const allErrors = [...foo, ...bar, ...baz]`